base/tuple.jl

```diff
@@ -660,16 +660,9 @@ prod(x::Tuple{}) = 1
 # than the general prod definition is available.
 prod(x::Tuple{Int, Vararg{Int}}) = *(x...)
 
-all(x::Tuple{}) = true
-all(x::Tuple{Bool}) = x[1]
-all(x::Tuple{Bool, Bool}) = x[1]&x[2]
-all(x::Tuple{Bool, Bool, Bool}) = x[1]&x[2]&x[3]
-# use generic reductions for the rest
-
-any(x::Tuple{}) = false
-any(x::Tuple{Bool}) = x[1]
-any(x::Tuple{Bool, Bool}) = x[1]|x[2]
-any(x::Tuple{Bool, Bool, Bool}) = x[1]|x[2]|x[3]
+all(x::NTuple{N,Bool}) where N = N <= 32 ? reduce(&, x; init = true) : _all(x, :)
+
+any(x::NTuple{N,Bool}) where N = N <= 32 ? reduce(|, x; init = false) : _any(x, :)
```

Comment:
Given that the tuple is homogeneous, there's no need for recursion, we can handle inputs of all sizes with a loop. This results in the same LLVM code for a length-32 tuple as with your variant with `reduce`:

```suggestion
function _all_bool(x::Tuple{Vararg{Bool}})
    @_terminates_locally_meta
    r = true
    for i ∈ eachindex(x)
        r &= getfield(x, i, false)
    end
    r
end

function _any_bool(x::Tuple{Vararg{Bool}})
    @_terminates_locally_meta
    r = false
    for i ∈ eachindex(x)
        r |= getfield(x, i, false)
    end
    r
end

any(x::Tuple{Vararg{Bool}}) = _any_bool(x)
all(x::Tuple{Vararg{Bool}}) = _all_bool(x)
all(x::Tuple{Bool}) = _all_bool(x)  # disambiguate
```